das/dastree/dastree.go

```diff
@@ -23,7 +23,13 @@ type node struct {
 	size uint32
 }
 
-func RecordHash(record func(bytes32, []byte), preimage ...[]byte) bytes32 {
+// RecordHash chunks the preimage into 64kB bins and generates a recursive hash tree,
+// calling the caller-supplied record function for each hash/preimage pair created in
+// building the tree structure.
+//
+// Notes
+//     1. RecordHash must only return an error if record returns an error.
+func RecordHash(record func(bytes32, []byte) error, preimage ...[]byte) (bytes32, error) {
```

Comment:
You don't have to do this, but the `record` closure doesn't have to return an `error`. The closure could instead save all the hashes, which you'd then post to IPFS after `RecordHash` returns.

Note that `record`, unlike `oracle`, isn't actually needed when constructing the DasTree hash. The same is not true for `Content`, which really does require a fallible closure.

I'd consider doing it this way since you'd be able to go back to the simpler function we originally had (and it would simplify the call sites). Not necessary though especially if streaming is a concern. Although if you did do it, you could do all your IPFS stores in parallel, which would be cool. So tradeoffs. If it turns out stores are expensive, you should probs switch to this model

Reply:
Great suggestion, thanks. Made this change.